i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,180 @@
+        which will be applied/added to dot product, used to mask padded key positions out
+        """
+        output_tensor = self.layernorm(input_tensor)  # [B,T,F]
+
+        time_dim_size = output_tensor.shape[1]
+        batch_dim_size = output_tensor.shape[0]
+
+        # attention mask
+        inv_sequence_mask = compat.logical_not(sequence_mask)  # [B, T]
+        mask = (
+            torch.zeros_like(inv_sequence_mask, dtype=input_tensor.dtype)
+            .masked_fill(inv_sequence_mask, float("-inf"))
+            .view(batch_dim_size, 1, 1, time_dim_size)
+            .expand(-1, self.num_heads, -1, -1)
+        )  # [B, #heads, 1, T']
+
+        # query, key and value sequences
+        bias_k, bias_q, bias_v = self.in_proj_bias.chunk(3)
+
+        query_seq = F.linear(output_tensor, self.q_proj_weight, bias_q)  # [B, T, #heads * F']
+        key_seq = F.linear(output_tensor, self.k_proj_weight, bias_k)
+        value_seq = F.linear(output_tensor, self.v_proj_weight, bias_v)
+
+        q1 = query_seq.view(batch_dim_size, -1, self.num_heads, self.embed_dim_per_head).transpose(
+            1, 2
+        )  # [B, #heads, T, F']
+        k_t = key_seq.view(batch_dim_size, -1, self.num_heads, self.embed_dim_per_head).permute(
+            0, 2, 3, 1
+        )  # [B, #heads, F', T']
+        # attention between query and key sequences
+        attn1 = torch.matmul(q1, k_t)  # [B, #heads, T, T']i
+
+        if self.rel_pos_clip:
+            q2 = (
+                query_seq.transpose(0, 1)
+                .contiguous()
+                .view(time_dim_size, batch_dim_size * self.num_heads, self.embed_dim_per_head)
+            )  # [T, B*#heads, F']
+
+            range_vec_q = torch.arange(time_dim_size, device=input_tensor.device)
+            range_vec_k = torch.arange(time_dim_size, device=input_tensor.device)
+
+            distance_mat = range_vec_k[None, :] - range_vec_q[:, None]
+            distance_mat_clipped = torch.clamp(distance_mat, -self.rel_pos_clip, self.rel_pos_clip)
+
+            final_mat = distance_mat_clipped + self.rel_pos_clip
+            # relative positional embeddings
+            rel_pos_embeddings = self.rel_pos_embeddings[final_mat]  # [T, T', F']
+
+            # attention between query sequence and relative positional embeddings
+            attn2 = torch.matmul(q2, rel_pos_embeddings.transpose(1, 2)).transpose(0, 1)  # [B*#heads, T, T']
+            attn2 = attn2.contiguous().view(
+                batch_dim_size, self.num_heads, time_dim_size, time_dim_size
+            )  # [B, #heads, T, T']
+
+            attn = (attn1 + attn2 + mask) * (math.sqrt(1.0 / float(self.embed_dim_per_head)))  # [B, #heads, T, T']
+        else:
+            attn = (attn1 + mask) * (math.sqrt(1.0 / float(self.embed_dim_per_head)))  # [B, #heads, T, T']
+
+        # softmax and dropout
+        attn_output_weights = F.dropout(
+            F.softmax(attn, dim=-1), p=self.att_weights_dropout, training=self.training
```

Comment:
For the tensorflow backend, is the broadcast also applied to the attention weight dropout?

Reply:
That depends on how the user configures it. The RETURNN TF backend can do whatever you want. But I think most people did not apply the broadcast here.

Reply:
Okay, then it is fine